docs/source/deployment-experience.md

```diff
@@ -0,0 +1,135 @@
+
+We recommend:
+
+* Use a central cluster, shared by all instruments.
+* Use authentication between the MongoDB client and server.
+
+### Message Bus
+
+We use a message to "publish" Bluesky documents from the RunEngine so
+that they are available for consumers in other processes or on
+other machines on the network.
+
+To start, we made good use of 0MQ as a simple message bus. However, we
+are moving to use Kafka, a much more fully-featured and robust solution.
+The Bluesky document publishers and consumers and fairly agnostic to
+the transport technology used, so switching is easy.
+
+Finally, we current subscribe the database-insertion code (`db.insert`)
+_directly_ to the RunEngine. We have begun to move that step downstream
+of a message bus, i.e. switching from:
+
+```
+RunEngine --> MongoDB
+```
+
+to
+
+```
+RunEngine --> Kafka --> MongoDB
+```
+
+We find that this improves performance because Kafka can accept data buffer it
+faster than MongoDB can insert it. We have yet to commit to this architecture
+at scale.
+
+We recommend:
+* Start with 0MQ to play because it is easy to set up.
+* Move to Kafka early, once you are relying on these systems.
+
+### Managing facility-specific code
+
+Where possible, we encourage facilities to push embellishments to Bluesky
+libraires so the benefits and maintenace can be shared. But each facility also
+accumulates utilies, bluesky plans, ophyd Devices, and other code that is
+specific to them. Some of this work may eventually be pushed upstream but is
+deployed locally at first to meet local deadlines. Other work may not be
+useful at other facilties and therefore not appropriate to contribute upstream.
+
+At NSLS-II, we have a catch-all library [nslsii](https://github.com/NSLS-II/nslsii).
+APS and LCLS have similar packages.
+
+We recommend:
+* Make a facility-specific package.
+* Make it public.
+
+### Managing instrument configuration
+
+Each instrument needs Python code to:
+
+* Define instrument-specific ophyd Device classes
+* Instantiate ophyd Devices
+* Define instrument-specific plans
+* Subscribe instrument-specific convenience
+* Customize the user experience
+
+Setting up an environment for collection is too complex for a declarative configuration
+file (like a YAML file) and requires executing Python code.
+
+The initial target of Bluesky development was a command-line tool. We chose the
+IPython application, we used the "profiles" feature of IPython to run
+instrument-specific code and populate the user namespace at IPython start-up
+with variables representing ophyd Device, bluesky plans, and so on.
+We currently have at least one GitHub repository per instrument. These are public
+under separate GitHub organization. The repositories are all named `profile_collection`.
+
+Over time, these large IPython profiles have become unwieldy and difficult to
+maintain. We recommend moving at least class and plan definitions to a normal
```

Comment:
This has resulted in conflicts with the centralized python environment, because instrument packages cannot be straightforwardly installed in development mode without clever hacks. More thought is needed about how this should work, from a beamline developer's perspective.

Reply:
`python setup.py develop` or `pip install -ve .` does two things is the setuptools world:

1. run `python setup.py dist_info` so that the right meta-data is generated
2. adds a `.pth` file where ever you are installing to add the path to where the source is to the PYTHONPATH at start up

For security reasons (I think), `.pth` files are only used when they are in the "standard" search path not in a folder added by `PYTHONPATH`, those paths are all read only so the overlay trick does not work here.  However, you can still do these two things manually or run `pip install ....` everytime you make a change (which if you wrap it up in a helper script that then restarts bsui might not be that bad).

Reply:
Or, use a starter shell script to include the instrument package as the interactive session is started.  Or advise how to use `sys.path.append()` to identify additional packages to be found.

Reply:
My point is that this recommendation to move to packages is made, but from my perspective the details haven't really been worked out in terms of actually making it easier to maintain and deploy in concert with the central conda env, which is also being recommended. I don't think I'd be able to convince someone who already had code in an IPython profile that they should switch to packaged code, as their life will get harder and more mysterious, and doubly so if they don't already understand the intricacies of packaging. The fact that we are discussing `python setup.py dist_info` and the locations where `.pth` files will and will not be read is a good indicator that more work is required to make this a turnkey solution for most people.